photosynthesis_metrics/msid.py

```diff
@@ -0,0 +1,420 @@
+r"""Implemetation of Multi-scale Evaluation metric, based on paper
+ https://arxiv.org/abs/1905.11141 and author's repository https://github.com/xgfs/msid
+"""
+from functools import partial
+from typing import List
+
+import torch
+import torch.nn.functional as f
+import torch.nn as nn
+import numpy as np
+
+import scipy.sparse as sps
+from scipy.sparse import lil_matrix, diags, eye
+
+from .utils import _validate_input
+EPSILON = 1e-6
+NORMALIZATION = 1e6
+
+# ---- laplacian.py
+def np_euc_cdist(data):
+    dd = np.sum(data*data, axis=1)
+    dist = -2*np.dot(data, data.T)
+    dist += dd + dd[:, np.newaxis] 
+    np.fill_diagonal(dist, 0)
+    np.sqrt(dist, dist)
+    return dist
+
+
+def construct_graph_sparse(data, k):
+    n = len(data)
+    spmat = lil_matrix((n, n))
+    dd = np.sum(data*data, axis=1)
+    
+    for i in range(n):
+        dists = dd - 2*data[i, :].dot(data.T)
+        inds = np.argpartition(dists, k+1)[:k+1]
+        inds = inds[inds!=i]
+        spmat[i, inds] = 1
+            
+    return spmat.tocsr()
+
+
+def construct_graph_kgraph(data, k):
+    import pykgraph
```

Comment:
Moved import on top, but there is a problem with that library.
It should be [compiled from source ](https://github.com/aaalgo/kgraph#building-and-installation) and only used in `_construct_graph_kgraph` function. There is other function `_construct_graph_sparse` which is being called by default and produces more accurate result (according to the paper). This library may be needed for really big datasets (like Imagenet), but its unclear for now how faster `kgraph` method comparing to `sparse`. 
For now I commented this line, because it's unclear if we should have such strong dependancy for just one line of code.

Reply:
Regarding the concerns, I would propose to get rid of this lib and the function that uses it util the moment when we actually need it. For now it seems like unnecessary complication.